caching/base.py

```diff
@@ -104,28 +112,47 @@ def __iter__(self):
                 yield obj
             return
 
-        # Do the database query, cache it once we have all the objects.
-        iterator = self.iter_function()
+        # Use the special FETCH_BY_ID iterator if configured.
+        if config.FETCH_BY_ID and hasattr(self.queryset, 'fetch_by_id'):
+            iterator = self.queryset.fetch_by_id
 
+        # No cached results. Do the database query, and cache it once we have
+        # all the objects.
         to_cache = []
-        try:
-            while True:
-                obj = next(iterator)
-                obj.from_cache = False
-                to_cache.append(obj)
-                yield obj
-        except StopIteration:
-            if to_cache or config.CACHE_EMPTY_QUERYSETS:
-                self.cache_objects(to_cache, query_key)
-            raise
+        for obj in iterator():
+            obj.from_cache = False
+            to_cache.append(obj)
+            yield obj
+        if to_cache or config.CACHE_EMPTY_QUERYSETS:
+            self.cache_objects(to_cache, query_key)
 
-    def cache_objects(self, objects, query_key):
-        """Cache query_key => objects, then update the flush lists."""
-        log.debug('query_key: %s' % query_key)
-        query_flush = flush_key(self.query_string)
-        log.debug('query_flush: %s' % query_flush)
-        cache.add(query_key, objects, timeout=self.timeout)
-        invalidator.cache_objects(self.model, objects, query_key, query_flush)
+
+class CacheMachine(CacheInternalCommonMixin):
```

Comment:
Not entirely sure this will work, but I wonder if `CacheInternalCommonMixin`, `CacheMachine`, and `CachingModelIterable` could all be collapsed into a single class (e.g., by setting `ModelIterable = object` when it fails to import). Definitely ignore this comment if that would complicate rather than simplify things!

Reply:
I tried doing that, but gave up early because of issues that I can't remember. But some of those problems might have been 2 small issues in the tests which I since identified, so let me try again.